lib/Dialect/LLHD/Transforms/Mem2Reg.cpp

```diff
@@ -669,32 +758,76 @@ LogicalResult Promoter::promote() {
 /// Identify any promotable slots probed or driven under the current region.
 void Promoter::findPromotableSlots() {
   SmallPtrSet<Value, 8> seenSlots;
+  SmallPtrSet<Operation *, 8> checkedUsers;
+  SmallVector<Operation *, 8> userWorklist;
+
   region.walk([&](Operation *op) {
     for (auto operand : op->getOperands()) {
       if (!seenSlots.insert(operand).second)
         continue;
 
-      // Ensure the slot is not used in any way we cannot reason about.
+      // We can only promote probes and drives on a locally-defined signal.
+      // Other signals, such as the ones brought into a module through a port,
+      // have an unknown aliasing relationship with the other ports.
       if (!operand.getDefiningOp<llhd::SignalOp>())
         continue;
-      if (!llvm::all_of(operand.getUsers(), [&](auto *user) {
-            // We don't support nested probes and drives.
-            if (region.isProperAncestor(user->getParentRegion()))
+
+      // Ensure the slot is not used in any way we cannot reason about.
+      auto checkUser = [&](Operation *user) -> bool {
+        // We don't support nested probes and drives.
+        if (region.isProperAncestor(user->getParentRegion()))
+          return false;
+        // Ignore uses outside of the region.
+        if (user->getParentRegion() != &region)
+          return true;
+        // Projection operations are okay as long as they are in the same block
+        // as any of their users.
+        if (isa<SigArrayGetOp>(user)) {
+          for (auto *projectionUser : user->getUsers()) {
+            if (projectionUser->getBlock() != user->getBlock())
```

Comment:
Isn't that too restrictive for many practical cases due to CSE of projection ops? (not blocking)

Reply:
Yes that's a good point 👍. We should be able to relax this by only requiring that the projection op dominates its users. That would also allow projections in dominating predecessor blocks and in the module to be covered, but still allow us to not reason about signals references through block arguments (that'll be a mighty headache).